latex/DPIA/primitives.tex

```diff
@@ -1,96 +1,66 @@
 \begin{figure}
     \footnotesize
-    \begin{minipage}{1.0\linewidth} \begin{tabular*}{\linewidth}{>{$}l<{$}@{\hspace{0.4em}}>{$}c<{$}>{$}l<{$}}
+    \begin{minipage}{1.0\linewidth}
+      \begin{tabular*}{\linewidth}{>{$}l<{$}@{\hspace{0.4em}}>{$}c<{$}>{$}l<{$}}
           \underline{\ell}&:&(\dt: \datatype)
             \to \tyexp{\dt}{\readacc}\\
 
-          \prim{negate}&:&(\dt: \datatype)
-            \to \tyexp{\dt}{\readacc}
-            \to \tyexp{\dt}{\readacc}\\
-
-          (+,*,/,-)   &:&(\dt: \datatype)
-            \to \tyexp{\dt}{\readacc}
-            \to \tyexp{\dt}{\readacc}
-            \to \tyexp{\dt}{\readacc}\\
-
-          \medskip\\
+          \prim{negate}&:&
+            \tyexp{\basic}{\readacc}
+              \to \tyexp{\basic}{\readacc}\\
```

Comment:
where is the dependent function gone?

Reply:
It shouldn't be dependent on the data type. `basic` covers vector and scalar types. Arrays and Tuples should not be allowed to be passed to `negate`.

Reply:
it should be dependent on the basic data type.

Reply:
in the formalism basic data type is a data type (our only "smallest" data type in fact). A dependent function takes as parameter something of a kind. If we say that the primitive depends on `basic` then the dependent function would be applied with a value of `basic`.

As an example, that would be like saying we have a dependent function that expects an `int`.

Reply:
are we not using dependent functions for primitives working on scalars?

Reply:
ah, seems like that was also removed recently

Reply:
In the implementation you may interpret it that way, but in the formalism (so far), we are hiding the differences between scalar (and vector) types but just talking about the artificial type `basic`.
As soon as we have to differ between vector and scalar types we may want to be more specific, but I do not see a reason for differentiating between `double`, `float` and `int` for example.

Reply:
should we not properly express that we have `scalar types <= basic types <= data types` in our formalism?

Reply:
it would be more like `scalar types <= basic types : data type`

You could do it but we didn't so far and I do not see a need for it yet. We would have to introduce a new kind `basic type` which in turn has kind `data type` or we would introduce subtyping for data types, something that I would like to avoid if possible.

Reply:
Should we open an improvement issue about this? We are showing primitives like cast and asVector that only work with scalar types, and yet they are dependent on any data type. Also, I thought that we wanted the formalism to match the implementation as much as possible. Now we have dependent functions in the implementation that are not dependent in the formalism. @michel-steuwer

Reply:
The vector types should definitely be modeled explicitly. The `asVector` primitive is a very good example for that. `cast` doesn't bother me too much. There it would be enough to say that we have different basic types. By keeping the scalar types all grouped under `basic` we do not have to cover different cases for different scalar types.